File: backend/internal/entity/ai_agent.go
package entity

import (
	"database/sql/driver"
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type AIAgent struct {
	ID               uuid.UUID                 `json:"id" db:"id"`
	TenantID         uuid.UUID                 `json:"tenant_id" db:"tenant_id" validate:"required"`
	Name             string                    `json:"name" db:"name" validate:"required"`
	Provider         string                    `json:"provider" db:"provider" validate:"required, oneof=claude openai google"`
	Model            string                    `json:"model" db:"model" validate:"required"`
	SystemPrompt     string                    `json:"system_prompt" db:"system_prompt"`
	Persona          *AiPersonaConfig          `json:"persona" db:"persona"`
	Safety           *AiSafetyConfig           `json:"safety" db:"safety"`
	Guardrails       *AiSafetyGuardrailsConfig `json:"guardrails" db:"guardrails"`
	Temperature      float64                   `json:"temperature" db:"temperature"`
	MaxTokens        int                       `json:"max_tokens" db:"max_tokens"`
	AutoReplyEnabled bool                      `json:"auto_reply_enabled" db:"auto_reply_enabled"`
	WorkingHours     *WorkingHours             `json:"working_hours" db:"working_hours"`
	IsActive         bool                      `json:"is_active" db:"is_active"`
	CreatedAt        time.Time                 `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time                 `json:"updated_at" db:"updated_at"`
}

type AiPersonaConfig map[string]interface{}
type AiSafetyConfig map[string]interface{}
type AiSafetyGuardrailsConfig map[string]interface{}
type WorkingHours map[string]interface{}

func (a AiPersonaConfig) Value() (driver.Value, error) {
	if a == nil {
		return nil, nil
	}
	return json.Marshal(a)
}
func (a *AiPersonaConfig) Scan(value interface{}) error {
	if value == nil {
		*a = nil
		return nil
	}
	bytes, ok := value.([]byte)
	if !ok {
		return nil
	}
	return json.Unmarshal(bytes, a)
}
func (a AiSafetyConfig) Value() (driver.Value, error) {
	if a == nil {
		return nil, nil
	}
	return json.Marshal(a)
}
func (a *AiSafetyConfig) Scan(value interface{}) error {
	if value == nil {
		*a = nil
		return nil
	}
	bytes, ok := value.([]byte)
	if !ok {
		return nil
	}
	return json.Unmarshal(bytes, a)
}
func (a AiSafetyGuardrailsConfig) Value() (driver.Value, error) {
	if a == nil {
		return nil, nil
	}
	return json.Marshal(a)
}
func (a *AiSafetyGuardrailsConfig) Scan(value interface{}) error {
	if value == nil {
		*a = nil
		return nil
	}
	bytes, ok := value.([]byte)
	if !ok {
		return nil
	}
	return json.Unmarshal(bytes, a)
}
func (w WorkingHours) Value() (driver.Value, error) {
	if w == nil {
		return nil, nil
	}
	return json.Marshal(w)
}
func (w *WorkingHours) Scan(value interface{}) error {
	if value == nil {
		*w = nil
		return nil
	}
	bytes, ok := value.([]byte)
	if !ok {
		return nil
	}
	return json.Unmarshal(bytes, w)
}

File: frontend/src/lib/server/ai-agent.ts
import { BaseHandler } from './base';

export interface AiPersona {
	agent_name?: string;
	soul?: string;
	tone?: string;
	gender?: string;
	greeting?: string;
	language?: string;
	[k: string]: unknown;
}
export interface AiSafety {
	handoff_to_human?: boolean;
	sensitive_topics?: string[];
	[k: string]: unknown;
}
export interface AiGuardrails {
	rules?: string[];
	[k: string]: unknown;
}

export interface AIAgent {
	id: string;
	name: string;
	provider: 'claude' | 'openai' | 'google';
	model: string;
	system_prompt: string;
	persona?: AiPersona | null;
	safety?: AiSafety | null;
	guardrails?: AiGuardrails | null;
	temperature: number;
	max_tokens: number;
	auto_reply_enabled: boolean;
	is_active: boolean;
	created_at?: string;
	updated_at?: string;
}

export interface AIAgentPayload {
	name: string;
	provider: string;
	model?: string;
	system_prompt: string;
	persona?: AiPersona;
	safety?: AiSafety;
	guardrails?: AiGuardrails;
	temperature?: number;
	max_tokens?: number;
	auto_reply_enabled?: boolean;
	is_active?: boolean;
}

export interface AIKnowledge {
	id: string;
	ai_agent_id: string;
	title: string;
	content: string;
	token_count: number;
}
export interface AIPlaybook {
	id: string;
	name: string;
	keywords: string[];
	instruction: string;
	asset_ids: string[];
	priority: number;
	is_active: boolean;
}
export interface AIAsset {
	id: string;
	name: string;
	type: 'image' | 'video' | 'document';
	storage_key: string;
	mime_type?: { String: string; Valid: boolean } | string | null;
	size?: unknown;
	description?: { String: string; Valid: boolean } | string | null;
}

type ApiRes<T> = { success: boolean; message?: string; data?: T };

export class AIAgentHandler extends BaseHandler {
	private base = '/ai-agents/protected';

	// ---- Agents ----
	async list(): Promise<AIAgent[]> {
		const res = await this.api.authRequest<AIAgent[]>('GET', `${this.base}/`);
		return res.success && res.data ? res.data : [];
	}
	async get(id: string): Promise<AIAgent | null> {
		const res = await this.api.authRequest<AIAgent>('GET', `${this.base}/${id}`);
		return res.success ? (res.data ?? null) : null;
	}
	async create(payload: AIAgentPayload): Promise<ApiRes<AIAgent>> {
		return this.api.authRequest<AIAgent>('POST', `${this.base}/`, payload);
	}
	async update(id: string, payload: AIAgentPayload): Promise<ApiRes<AIAgent>> {
		return this.api.authRequest<AIAgent>('PUT', `${this.base}/${id}`, payload);
	}
	async remove(id: string): Promise<ApiRes<null>> {
		return this.api.authRequest<null>('DELETE', `${this.base}/${id}`);
	}

	// ---- Nested resources (knowledge | playbooks | assets) ----
	async listResource<T>(agentId: string, resource: string): Promise<T[]> {
		const res = await this.api.authRequest<T[]>('GET', `${this.base}/${agentId}/${resource}/`);
		return res.success && res.data ? res.data : [];
	}
	async createResource<T>(agentId: string, resource: string, payload: unknown): Promise<ApiRes<T>> {
		return this.api.authRequest<T>('POST', `${this.base}/${agentId}/${resource}/`, payload);
	}
	async updateResource<T>(
		agentId: string,
		resource: string,
		childId: string,
		payload: unknown
	): Promise<ApiRes<T>> {
		return this.api.authRequest<T>('PUT', `${this.base}/${agentId}/${resource}/${childId}`, payload);
	}
	async deleteResource(agentId: string, resource: string, childId: string): Promise<ApiRes<null>> {
		return this.api.authRequest<null>('DELETE', `${this.base}/${agentId}/${resource}/${childId}`);
	}
}
